### PR TITLE
Introduce RxJS to standardise and simplify Observables

### PR DESCRIPTION
**Pull request by max-vogler**
**This PR includes changes from https://github.com/GoogleChrome/audion/pull/111 as diffbase.**

Introduce [RxJS](https://rxjs.dev/) to replace custom Observables. Using a standardised, well-tested library like RxJS brings several benefits:
 * Code edge cases are well tested, e.g. currently [WholeGraphButton](https://github.com/GoogleChrome/audion/blob/fda9b9939808d84ccd3b0186773823ed96ff98b1/src/panel/components/WholeGraphButton.js#L16) stops emitting to an initial Observer when another Observer subscribes, [filter](https://github.com/GoogleChrome/audion/blob/fda9b9939808d84ccd3b0186773823ed96ff98b1/src/utils/Observer.js#L76) silently ignores errors and Observable completion and other edge cases are largely undefined and untested.
 * RxJS Observables can be composed in a declarative manner through `pipe`, which makes code more readable.
 * RxJS uses well-understood names and paradigms.

This increases the bundle size from 179k to 188k. Bundle size will decrease slightly in the future, when custom Observable implementation is removed.

Diff for refactored WholeGraphButton logic:

```diff
- export class WholeGraphButton {
-   constructor() {
-     // ...
-     this._clickNext = null;
-     this._clickObserver = new Observer((onNext) => {
-       this._clickNext = onNext;
-       return () => {
-         this._clickNext = null;
-       };
-     });
-     this.click = this.click.bind(this);
-   }
- 
-   observe(onNext, onComplete, onError) {
-     return this._clickObserver.observe(() => {
-       onNext({event: 'resizeView'});
-     });
-   }
- 
-   click() {
-     if (this._clickNext) {
-       this._clickNext();
-     }
-   }
- 
-   render() {
-     // ...
-     this.view.addEventListener('click', this.click);
-     // ...
-   }
- }
+ export class WholeGraphButton {
+   readonly click$ = fromEvent(this.view, 'click');
+   // ...
+ }
=
= // Usage:
- wholeGraphButton.observe(() => /* ... */);
+ wholeGraphButton.click$.subscribe(() => /* ... *./);
```

Diff for refactored panel worker logic:
```diff
- /** @type {Utils.Observer<PanelMessage>} */
- const receiver = observeMessageEvents(self);
+ const messages$ = fromEvent<PanelMessage>(self, 'message').pipe(
+   map((message) => message.data),
+ );
```

```diff
- let layoutOptions = {rankdir: 'LR'};
- const graphContextMessageFilterObserver = new Observer((onNext, ...args) => {
-   return receiver.observe((data) => {
-     if ('layoutOptions' in data) {
-       layoutOptions = data.layoutOptions;
-     } else {
-       onNext(data.graphContext);
-     }
-   }, ...args);
- });
+ const layoutOptions$: Observable<dagre.GraphLabel> = messages$.pipe(
+   filter((msg): msg is LayoutOptionsMessage => 'layoutOptions' in msg),
+   map((message) => message.layoutOptions),
+   startWith({rankdir: 'LR'}),
+ );
```
```diff
- const deserializedObserver = Observer.transform(
-   graphContextMessageFilterObserver,
-   deserializeGraphContext,
- );
- 
- const receiverThrottle = Observer.throttle(deserializedObserver, {
-   key: (message) => message.id,
- });
- 
- const updateGraph = Observer.transform(receiverThrottle, (context) => {
-   if (context.context && context.graph) {
-     context.graph.setGraph(layoutOptions);
-     dagre.layout(context.graph);
-   }
-   return context;
- });
- 
- const serializedGraph = Observer.transform(updateGraph, serializeGraphContext);
- 
- postObservations(serializedGraph, self);
+ messages$
+   .pipe(
+     filter((msg): msg is GraphContextMessage => 'graphContext' in msg),
+     throttleTime(16),
+     map((message) => deserializeGraphContext(message.graphContext)),
+     withLatestFrom(layoutOptions$),
+     map(([context, layoutOptions]) => {
+       if (context.context && context.graph) {
+         context.graph.setGraph(layoutOptions);
+         dagre.layout(context.graph);
+       }
+       return context;
+     }),
+     map(serializeGraphContext),
+   )
+   .subscribe((context) => {
+     self.postMessage(context);
+   });

### CODE REVIEW

**Comment by hoch** [comment]
Thanks! I'll review this once we have #111 merged.

**Comment by max-vogler** [comment]
Merged #111 and rebased this PR. PTAL.

File: src/devtools/Types.ts [hunk 1]
```diff
@@ -17,18 +40,6 @@
  * @property {object} graph
  */
 
-/**
```
Comment (hoch) [inline src/devtools/Types.ts:20]: Wouldn't this break the JSDoc export? Don't we need to keep the annotation with the change above?

File: src/devtools/WebAudioGraphIntegrator.ts [hunk 2]
```diff
@@ -99,7 +104,10 @@ export class WebAudioGraphIntegrator extends Observer<Audion.GraphContext> {
         context: contextCreated.context,
         nodes: {},
         params: {},
-        graph,
+        // TODO: dagre's graphlib typings are inacurate, which is why we use
```
Comment (hoch) [inline src/devtools/WebAudioGraphIntegrator.ts:107]: typo: inaccurate

File: src/devtools/deserializeGraphContext.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  graph: any;
+}
+
+export function deserializeGraphContext(
+  graphContext: SerializedGraphContext,
+): Audion.GraphContext {
+  if (graphContext.graph) {
+    return {
+      ...graphContext,
+      // TODO: dagre's graphlib typings are inacurate, which is why we use
```
Comment (hoch) [inline src/devtools/deserializeGraphContext.ts:14]: typo: inacurate

File: src/panel/worker.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+messages$
+  .pipe(
+    filter((msg): msg is GraphContextMessage => 'graphContext' in msg),
+    throttleTime(16),
+    map((message) => deserializeGraphContext(message.graphContext)),
+    withLatestFrom(layoutOptions$),
+    map(([context, layoutOptions]) => {
+      if (context.context && context.graph) {
+        context.graph.setGraph(layoutOptions);
+        // TODO: dagre's graphlib typings are inacurate, which is why we use
```
Comment (hoch) [inline src/panel/worker.ts:47]: typo: inacurate

**Review by hoch — COMMENTED** [review]
What's the level of maintenance of RxJS? How big is it? The actual logic change in this PR is pretty small, so I am not seeing the notable benefit of using RxJS.

**Comment by hoch** [comment]
You can ignore my comment above. The PR description explains everything I needed to know. LGTM.

**Review by hoch — APPROVED** [review]

File: src/devtools/Types.ts (hunk 1 above)
Comment (max-vogler) [reply src/devtools/Types.ts:20]: Good point. I was unaware that JSDoc is actively used in this project and indeed, [JSDoc is incompatible with TypeScript](https://stackoverflow.com/questions/51720894/is-there-any-way-to-use-jsdoc-with-ts-files-maybe-transpile-with-babel-the). 

Can you explain the purpose/intentions of JSDoc here? The docs seems to only be executed as side-effect of tests and are not hosted anywhere.

**Comment by max-vogler** [comment]
> What's the level of maintenance of RxJS?

I've been using RxJS for about a year in https://github.com/google/grr and encountered no maintenance churn. The API seems stable, conforms to semver, and upgrading from v6 to v7 was unproblematic. For the 3-year old v6, rxjs still pushed a new release this year, so the maintenance cycles seem acceptable.

**Pull request merged**